packages/core/src/utils/SIWXUtil.ts

```diff
@@ -141,8 +141,16 @@ export const SIWXUtil = {
   },
   async cancelSignMessage() {
     try {
-      await ConnectionController.disconnect()
+      const siwx = this.getSIWX()
+      const required = siwx?.getRequired()
+
+      if (required) {
+        await ConnectionController.disconnect()
+      }
```

Comment:
maybe would this be the case? I guess this is going to always close the modal
```suggestion
      if (required) {
        await ConnectionController.disconnect()
        RouterController.reset('Connect')
      } else {
        ModalController.close()
      }
```